components/SearchBar.tsx

```diff
@@ -0,0 +1,61 @@
+/* This file is configured based on the following Algolia demo:
+  https://codesandbox.io/s/exciting-gwen-jr5kj?file=/src/Autocomplete.tsx
+
+  The file groups all search bar components with one another. The search bar
+  consists of the SearchInput where users write their query, the SearchPanel
+  which displays the autocomplete options, and the autocomplete hook which makes
+  algolia queries.
+*/
+
+import { useState } from "react";
+import {
+  AutocompleteOptions,
+  AutocompleteState,
+} from "@algolia/autocomplete-core";
+import { Hit } from "@algolia/client-search";
+import { Book } from "@prisma/client";
+import { CourseWithDept } from "@lib/services/course";
+import { SearchInput } from "./SearchInput";
+import { SearchPanel, SearchPanelProps } from "./SearchPanel";
+import { useAutocomplete } from "@lib/hooks/autocomplete";
+
+export type AutocompleteItem = Hit<{
+  type: string;
+  entry: Book | CourseWithDept;
+}>;
```

Comment:
Consider doing this instead:
```typescript
export type AutocompleteItem = Hit<
  { type: "book"; entry: Book } | { type: "course"; entry: CourseWithDept }
>;
```

It will help avoid a bunch of unsafe type casts and prevent the unsafe use of magic strings ("book" and "course") across the frontend.